SUS-4611 | backport MemcachedPeclBagOStuff from MW 1.20

macbre opened the pull request:
https://wikia-inc.atlassian.net/browse/SUS-4611

This pull request includes a [backport from MW 1.20 that introduces `memcached` extension based client](https://github.com/wikimedia/mediawiki/commit/3c62077fe28778eb41bde7549fbae402c8be0ef7#diff-86946fda95cb93f7a1deb64d78e6c661). This will replace an old, Perl-based memcache client that is written in PHP.

I've also improved caching and errors handling in `VideoHandlers` classes that control video uploads from 3rd party providers. 

## Performance

In terms of performance there's no difference (tested on sandbox) in overall response time (median value for 100 requests difference by less than 1%). Profiling shows that the old client uses ~17% of the response time, while the new one - ~13%.

The new client is way more reliable when using a new localisation cache - see #15203. The code is much simpler (there's no sockets communication handling and results parsing in PHP code).

## Deployment

This PR **does not switch the memcache client on production nor on preview / verify**. For now it will only be enabled on sandboxes and devboxes.

Due to incompatibility of serialization of non-scalar types between these two clients we'd need to introduce a key prefix for a new client (that's a part of this PR). In order to avoid a flood of cache misses after production deploy we need to start enabling the new client on a small fraction of production requests first.

Commit:
SUS-4611 | backport MemcachedPeclBagOStuff from MW 1.20

https://github.com/wikimedia/mediawiki/commit/3c62077fe28778eb41bde7549fbae402c8be0ef7#diff-86946fda95cb93f7a1deb64d78e6c661

macbre committed:
Backport fixes

mszabo-wikia commented on includes/objectcache/MemcachedPeclBagOStuff.php:
```diff
@@ -0,0 +1,200 @@
+				} else {
+					$msg = "Memcached error: $msg";
+				}
+				wfDebugLog( 'memcached-serious', $msg );
```
Let's update logging statements as needed

macbre committed:
Type hints

macbre committed:
MemcachedPeclBagOStuff: type hinting

macbre committed:
Improve logging from memcache clients

macbre committed:
MemcachedPeclBagOStuff: support binary protocol

This is required for e.g. the touch() method in older versions of
the Memcached client.

It still defaults to ASCII.

Backporting https://github.com/wikimedia/mediawiki/commit/ad7a968fb025fe3d90e527a6e8c327c72bb093b3 from MW 1.28

macbre committed:
CACHE_MEMCACHED: use MemcachedPeclBagOStuff client

macbre committed:
CACHE_MEMCACHED: disable binary protocol

macbre committed:
Get rid of ObjectCache::newMemcached

macbre committed:
Remove a lonely MWLibMemcached class

macbre committed:
Remove LibmemcachedBagOStuff - old work in progress stuff

macbre committed:
MemcachedPeclBagOStuff - add a prefix to keys to avoid problems with incompatible serialization between clients

macbre committed:
CACHE_MEMCACHED: keep using the OLD memcache client

macbre committed:
MemcachedPeclBagOStuff: add some profiling

macbre committed:
MemcachedPeclBagOStuff: Memoize duplicate memcache requests for the same key in the same request

macbre committed:
Merge branch 'dev' into SUS-4611-use-memcached-ext

macbre committed:
MemcachedPhpBagOStuff: methods cleanup

macbre committed:
Merge branch 'dev' into SUS-4611-use-memcached-ext

macbre committed:
SUS-4611: use a new memcache client on sandboxes and devboxes

bkoval commented on includes/objectcache/MemcachedBagOStuff.php:
```diff
@@ -0,0 +1,176 @@
+	 * TTLs higher than 30 days will be detected as absolute TTLs
+	 * (UNIX timestamps), and will result in the cache entry being
+	 * discarded immediately because the expiry is in the past.
+	 * Clamp expiries >30d at 30d, unless they're >=1e9 in which
```
*expires?

macbre committed:
Merge branch 'dev' into SUS-4611-use-memcached-ext

macbre committed:
VideoEmbedTool: log exceptions thrown by ApiWrapper

macbre committed:
Merge branch 'dev' into SUS-4611-use-memcached-ext

macbre committed:
Remove MemcachedStats - we no longer collect detailed per-key memcache statistics regarding hit-miss ratio

macbre committed:
MemcachedPeclBagOStuff: use the caller name instead of a normalized key when reporting memcache errors

macbre committed:
VET: make sure AJAX responses are not cached

macbre committed:
VideoFileUploader: improve errors handling

macbre committed:
VET: add some errors logging

macbre committed:
ApiWrapper: more pesimistic approach to API responses

macbre committed:
ApiWrapper::checkForResponseErrors - assert we can get thumbnail URL from video provider API response

macbre committed:
ApiWrapper: simplify caching logic
